File: StopWatch/app/src/main/java/com/thomaskuenneth/stopwatch/StopwatchActivity.java
package com.thomaskuenneth.stopwatch;

import android.app.Activity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.Timer;
import java.util.TimerTask;

public class StopwatchActivity extends Activity {

    private static final String KEY_DIFF = "diff";
    private static final String KEY_RUNNING = "running";

    private final DateFormat format;
    private Timer timer;
    private TimerTask timerTask;
    private TextView time;
    private Button startStop;
    private Button reset;
    private long started;
    private boolean isRunning;
    private long diff;

    public StopwatchActivity() {
        format = new SimpleDateFormat("HH:mm:ss:SSS",
                Locale.US);
        format.setTimeZone(TimeZone.getTimeZone("UTC"));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        time = findViewById(R.id.time);
        startStop = findViewById(R.id.start_stop);
        startStop.setOnClickListener(v -> {
            isRunning = !isRunning;
            if (isRunning) {
                scheduleAtFixedRate();
            } else {
                timerTask.cancel();
            }
            updateUI();
        });
        reset = findViewById(R.id.reset);
        reset.setOnClickListener(v -> clearTime());
        if (savedInstanceState != null) {
            getValuesFromBundle(savedInstanceState);
            setTime();
        } else {
            isRunning = false;
            clearTime();
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putLong(KEY_DIFF, diff);
        outState.putBoolean(KEY_RUNNING, isRunning);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        getValuesFromBundle(savedInstanceState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        timer = new Timer();
        updateUI();
        if (isRunning) {
            scheduleAtFixedRate();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        timer.cancel();
    }

    private void clearTime() {
        time.setText(R.string.cleared);
        diff = 0;
    }

    private void updateUI() {
        startStop.setText(isRunning ? R.string.stop : R.string.start);
        reset.setEnabled(!isRunning);
    }

    private void getValuesFromBundle(Bundle b) {
        diff = b.getLong(KEY_DIFF);
        isRunning = b.getBoolean(KEY_RUNNING);
    }

    private void setTime() {
        time.setText(format.format(new Date(diff)));
    }

    private void scheduleAtFixedRate() {
        started = System.currentTimeMillis() - diff;
        timerTask = new TimerTask() {
            @Override
            public void run() {
                diff = System.currentTimeMillis() - started;
                runOnUiThread(() -> setTime());
            }
        };
        timer.scheduleAtFixedRate(timerTask, 0, 200);
    }
}

File: xamarin/Stopwatch/Stopwatch/StopwatchActivity.cs
using Android.App;
using Android.OS;
using Java.Text;
using Java.Util;
using System;
using Android.Widget;
using Java.Lang;
using Android.Views;

namespace Stopwatch
{
    [Activity(Label = "@string/app_name", Exported = true, MainLauncher = true)]
    public class StopwatchActivity : Activity
    {
        const string KEY_DIFF = "diff";
        const string KEY_RUNNING = "running";

        readonly DateFormat format;
        Timer timer;
        TimerTask timerTask;
        TextView time;
        Button startStop;
        Button reset;
        long started;
        bool isRunning;
        long diff;

        public StopwatchActivity()
        {
            format = new SimpleDateFormat("HH:mm:ss:SSS",
                    Locale.Us)
            {
                TimeZone = Java.Util.TimeZone.GetTimeZone("UTC")
            };
        }

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.Main);
            time = FindViewById<TextView>(Resource.Id.time);
            startStop = FindViewById<Button>(Resource.Id.start_stop);
            startStop.SetOnClickListener(new MyOnClickListener(() =>
            {
                isRunning = !isRunning;
                if (isRunning)
                {
                    ScheduleAtFixedRate();
                }
                else
                {
                    timerTask.Cancel();
                }
                UpdateUI();
            }));
            reset = FindViewById<Button>(Resource.Id.reset);
            reset.SetOnClickListener(new MyOnClickListener(() => { ClearTime(); }));
            if (savedInstanceState != null)
            {
                GetValuesFromBundle(savedInstanceState);
                SetTime();
            }
            else
            {
                isRunning = false;
                ClearTime();
            }
        }

        protected override void OnSaveInstanceState(Bundle outState)
        {
            base.OnSaveInstanceState(outState);
            outState.PutLong(KEY_DIFF, diff);
            outState.PutBoolean(KEY_RUNNING, isRunning);
        }

        protected override void OnRestoreInstanceState(Bundle savedInstanceState)
        {
            base.OnRestoreInstanceState(savedInstanceState);
            GetValuesFromBundle(savedInstanceState);
        }

        protected override void OnResume()
        {
            base.OnResume();
            timer = new Timer();
            UpdateUI();
            if (isRunning)
            {
                ScheduleAtFixedRate();
            }
        }

        protected override void OnPause()
        {
            base.OnPause();
            timer.Cancel();
        }

        private void ClearTime()
        {
            time.SetText(Resource.String.cleared);
            diff = 0;
        }

        private void UpdateUI()
        {
            startStop.SetText(isRunning ? Resource.String.stop : Resource.String.start);
            reset.Enabled = !isRunning;
        }

        private void GetValuesFromBundle(Bundle b)
        {
            diff = b.GetLong(KEY_DIFF);
            isRunning = b.GetBoolean(KEY_RUNNING);
        }

        private void SetTime()
        {
            time.Text = format.Format(new Date(diff));
        }

        private void ScheduleAtFixedRate()
        {
            started = JavaSystem.CurrentTimeMillis() - diff;
            timerTask = new MyTimerTask(() =>
            {
                diff = JavaSystem.CurrentTimeMillis() - started;
                RunOnUiThread(() =>
                {
                    SetTime();
                });
            });
            timer.ScheduleAtFixedRate(timerTask, 0, 200);
        }
    }

    public sealed class MyOnClickListener : Java.Lang.Object, View.IOnClickListener
    {
        readonly Action action;

        public MyOnClickListener(Action action)
        {
            this.action = action;
        }

        void View.IOnClickListener.OnClick(View v)
        {
            action();
        }
    }
}